File: src/src_core/fmg/core/types/event/MosaicEvent.java
package fmg.core.types.event;

import java.util.EventObject;

import fmg.core.mosaic.MosaicBase;
import fmg.core.mosaic.cells.BaseCell;

public class MosaicEvent extends EventObject {
   private static final long serialVersionUID = 1L;

   private MosaicEvent(MosaicBase source) {
      super(source);
   }

   public static class ClickEvent extends MosaicEvent {
      private static final long serialVersionUID = 1L;
      BaseCell cell;
      boolean leftClick;
      boolean down;

      public ClickEvent(MosaicBase source, BaseCell clickedCell, boolean leftClick, boolean down) {
         super(source);
         this.cell = clickedCell;
         this.leftClick  = leftClick;
         this.down = down;
      }

      public boolean isLeftClick() {
         return leftClick;
      }
      public boolean isDown() {
         return down;
      }
      public BaseCell getCell() {
         return cell;
      }
   }

   @Override
   public MosaicBase getSource() {
      return (MosaicBase)super.getSource();
   }
}

File: src/src_core/fmg/core/types/event/MosaicEvent.cs
using fmg.core.mosaic;
using fmg.core.mosaic.cells;
using fmg.core.mosaic.draw;

namespace fmg.core.types.Event
{

   public class MosaicEvent : System.EventArgs
   {

      public class ClickEventArgs : MosaicEvent
      {
         BaseCell cell;
         bool leftClick;
         bool down;

         public ClickEventArgs(BaseCell clickedCell, bool leftClick, bool down)
         {
            this.cell = clickedCell;
            this.leftClick = leftClick;
            this.down = down;
         }

         public bool isLeftClick()
         {
            return leftClick;
         }
         public bool isDown()
         {
            return down;
         }
         public BaseCell getCell()
         {
            return cell;
         }
      }

      public static MosaicBase<TPaintable> getSource<TPaintable>(object sender) where TPaintable : IPaintable {
         return sender as MosaicBase<TPaintable>;
      }
   }
}